File: lib/emerald/nodes/tag_statement.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'treetop'
require_relative 'node'

# A tag
class TagStatement < Node
  # http://w3c.github.io/html/syntax.html#void-elements
  VOID_TAGS = (
    "area base br col embed hr img input link " +
    "menuitem meta param source track wbr"
  ).split(/\s+/)

  def to_html(context)
    if void_tag?
      opening_tag(context)
    else
      opening_tag(context) +
        (
          if !body.empty?
            body.to_html(context)
          else
            ''
          end
        ) +
        closing_tag(context)
    end
  end

  def class_attribute
    unless classes.empty?
      class_names = classes
        .elements
        .map{ |c| c.name.text_value }
        .join(' ')

      " class=\"#{class_names}\""
    else
      ''
    end
  end

  def id_attribute
    unless identifier.empty?
      " id=\"#{identifier.name.text_value}\""
    else
      ''
    end
  end

  def opening_tag(context)
    "<#{tag.text_value}" +
      id_attribute +
      class_attribute +
      (
        if !attributes.empty?
          ' ' + attributes.to_html(context)
        else
          ''
        end
      ) + (
        if void_tag?
          ' />'
        else
          '>'
        end
      )
  end

  def void_tag?
    VOID_TAGS.include? tag.text_value
  end

  def closing_tag(_context)
    "</#{tag.text_value}>"
  end
end

File: cpp/src/nodes/tag_statement.cpp
#include "tag_statement.hpp"
#include <boost/algorithm/string/join.hpp>

TagStatement::TagStatement(
  std::string tag_name,
  std::string id,
  std::vector<std::string> classes,
  NodePtr body,
  NodePtr attributes,
  NodePtr nested
):
  tag_name(tag_name),
  id(id),
  classes(classes),
  body(body),
  attributes(attributes),
  nested(nested),
  self_closing(void_tags.find(tag_name) != void_tags.end())
{}

std::set<std::string> TagStatement::void_tags = {
  "area", "base", "br", "col", "embed", "hr", "img",
  "input", "link", "menuitem", "meta", "param", "source",
  "track", "wbr"
};

std::string TagStatement::to_html(Json &context) {
  if (self_closing) {
    return opening_tag(context);
  } else if (body) {
    return opening_tag(context) + body->to_html(context) + closing_tag();
  } else {
    return opening_tag(context) + nested->to_html(context) + closing_tag();
  }
}

std::string TagStatement::opening_tag(Json &context) const {
  return "<" + tag_name + id_attribute() + class_attribute() +
    (attributes ? " " + attributes->to_html(context) : "") +
    (self_closing ? " />" : ">");
}

std::string TagStatement::id_attribute() const {
  if (id.empty()) {
    return "";
  } else {
    return " id=\"" + id + "\"";
  }
}

std::string TagStatement::class_attribute() const {
  if (classes.size() == 0) {
    return "";
  } else {
    return " class=\"" + boost::algorithm::join(classes, " ") + "\"";
  }
}

std::string TagStatement::closing_tag() const {
  return "</" + tag_name + ">";
}
